spec/adapters/threema_adapter/outbound_spec.rb

```diff
@@ -24,29 +23,40 @@
 
   describe '::send_welcome_message!' do
     subject { -> { described_class.send_welcome_message!(contributor) } }
+
     it { should enqueue_job(described_class) }
+
     context 'contributor has no threema_id' do
       let(:contributor) { create(:contributor, threema_id: nil, email: nil) }
       it { should_not enqueue_job(described_class) }
     end
   end
 
   describe '#perform' do
-    subject { adapter.perform(text: message.text, recipient: message.recipient) }
+    subject { -> { adapter.perform(text: message.text, recipient: message.recipient) } }
 
     it 'sends the message' do
-      expect(threema).to receive(:send).with({ type: :text, threema_id: contributor.threema_id, text: message.text })
-
-      subject
+      expect_any_instance_of(Threema).to receive(:send).with(type: :text, threema_id: 'V5EA564T', text: message.text)
```

Comment:
It’s recommended not to use `expect_any_instance_of` and `allow_any_instance_of` as it’s an indicator for design smells.

Previously, we worked around this by stubbing the `new` method to always return an instance double. However, for some reason this failed as soon as I tried to use this approach in multiple test cases with the following error message:

> Double was originally created in one example but has leaked into another example and can no longer be used. rspec-mocks' doubles are designed to only last for one example, and you need to create a new one in each example you wish to use it for.

So I did what the message recommended and initialized a new double for each test case, which didn’t change anything though.

As stubbing `new` doesn’t really solve the underlying design issue (the reason why using `*_any_instance_of` is discouraged), I decided to use `expect_any_instance_of` instead, which doesn’t lead to leaks.

I’d still be interested in cleaner solutions, though.

Reply:
I also have read about how the use of `allow_any_instance_of` is discouraged. So, I would like to understand why this failure is happening and find a better solution.

I pulled down the code and did the same as you and ran into the same issue you did. I searched SO and found lots of ideas, one of them [rspec-mocks' doubles are designed to only last for one example](https://stackoverflow.com/questions/27088250/rspec-mocks-doubles-are-designed-to-only-last-for-one-example)](https://stackoverflow.com/questions/27088250/rspec-mocks-doubles-are-designed-to-only-last-for-one-example) suggested something that helped get to the bottom of the issue.

If you look at the [ThreemaAdapter](https://github.com/tactilenews/100eyes/blob/master/app/adapters/threema_adapter/outbound.rb#L7-L9), you can see that we are memoizing @threema_instance, so that when you call `self.class.threema_instance.send(...)` multiple times instead of using a new instance of `Threema`, it uses the memoized version. This might, or might not, be desirable for our application, but it makes testing it difficult.

It's not really clear to me what we gain from memoizing `Threema.new` in this way. The code was added while solving an issue that is unrelated and there is no explanation. See https://github.com/tactilenews/100eyes/pull/930/commits/8ad5c6f264c707262b90b5b4cecfca1ce6d55ba5.

I understand that what we want to achieve here is that for every threema recipient we don't set up a new instance of `Threema`, but I guess I'm doubtful if we gain a lot from that. Do you have any thoughts on that? @roschaefer you added this code, could you explain the benefit to me?

Reply:
Thanks for having a closer look into it! If we can find a closer solution to that without too much effort, I’m all in for it. If it turns out to be something that would take more time, I think we should balance that with the value we gain from it, especially as it’s something that we touch rarely.

Reply:
A quick solution would be to revert the change that memoizes the instance of `Threema`. I am not sure we gain a lot from memoizing here since creating an instance of the class can't be an expensive action, but of course I might be missing the larger purpose @roschaefer wanted to achieve by memomizing in the first place.